### ISSUE TITLE
SSUSessionConfirmed: review if/why Alice sends invalid identity size and/or certificate length

### ISSUE THREAD

**Issue by anonimal** [monero-project/kovri#432]
**By submitting this issue, I confirm the following:**

- I have read and understood the [contributor guide](https://github.com/monero-project/kovri/blob/master/doc/CONTRIBUTING.md).
- I have checked that the issue I am reporting can be replicated or that the feature I am suggesting is not present.
- I have checked opened or recently closed [pull requests](https://github.com/monero-project/kovri/pulls) for existing solutions/implementations to my issue/suggestion.

*Place an X inside the bracket to confirm*
- [X] I confirm.

---

First noticed on our OSX 10.12 machine:

```
2016:11:04|DBG   SSUServer: handling received data
2016:11:04|DBG   SSUServer: receiving data
2016:11:04|DBG   SSUServer: handling received packets
2016:11:04|DBG   SSUSession: [] [IPv4 address] --> 1456 bytes transferred, 4259 total bytes received
2016:11:04|ERR   IdentityEx: certificate length 23753 exceeds buffer length 83
Segmentation fault: 11
```

I was not running LLDB at the time but I've now attached all running kovri's that weren't previously attached.

If cert length exceeds buffer, we do return false but it's likely that, after the fact, we're not catching elsewhere (I'm not at my dev machine but can confirm later).  IIRC, this is the first time I've seen this issue (we really should have more unit-tests in this area though where possible).

Referencing #366

**Comment by anonimal** [comment]
Platform-agnostic (popped on both Ubuntu 32/64 and FreeBSD 64 machines), the segfault appears to be due to a null pointer source copy during copy assignment.

Note: interestingly, the session was sent from Alice in a private address space (10.0.2.2) and we blow up when parsing SessionConfirmed.

Side-note: we should ban IP's from private/reserved address spaces (this deserves it's own ticket).

Backtraces + logs:
- [Ubuntu i686 16.04](https://github.com/monero-project/kovri/files/601771/ubuntu-32_issue-432.txt)
- [Ubuntu amd64 16.04](https://github.com/monero-project/kovri/files/601772/ubuntu-64_issue-432.txt)
- [FreeBSD amd64 10.3](https://github.com/monero-project/kovri/files/601773/freebsd-64_issue-432.txt)

TL;DR:

Healthy packet example:
```
kovri::core::SSUPacketParser::ParseSessionConfirmed (this=0x7fffeb7fd770) at /home/anonimal/kovri/src/core/router/transports/ssu/packet.cc:768
768       kovri::core::IdentityEx identity;
(gdb) ins identity
$14 = {m_StandardIdentity = {
    public_key = "pO\005\334\377\177\000\000\300\351\177\353\377\177\000\000\000\000\000\000\000\000\000\000\064\000\000\000\000\000\000\000\300\324\177\353\377\177\000\000hU\005\334\377\177\000\000pO\005\334\377\177\000\000P\336\372\000\0
00\000\000\000pO\005\334\377\177\000\000\005\000\000\000\000\000\000\000\300O\006\334\377\177\000\000/\325\177\353\377\177\000\000\360\324\177\353\377\177\000\000\060\000\000\000\000\000\000\000\360\324\177\353\377\177\000\000\267\325\177\
353\377\177\000\000\000\325\177\353\377\177\000\000\270j\035\001\000\000\000\000\267\325\177\353\377\177\000\000\240\325\177\353\377\177\000\000 \325\177\353\377\177\000\000\356\202\035\001\000\000\000\000\240\325\177\353\377\177\000\000`\
326\177\353\377\177\000\000P\325\177\353\377\177\000\000"..., 
    signing_key = "\000\000\000\000\000\000\000\000`\326\177\353\377\177\000\000\220\325\177\353\377\177\000\000\020x&\001\000\000\000\000\004\000\000\000\000\000\000\000`\326\177\353\377\177\000\000\260\325\177\353\377\177\000\000`\326\17
7\353\377\177\000\000\300\325\177\353\377\177\000\000\366w&\001\000\000\000\000\320\325\177\353\377\177\000\000`\326\177\353\377\177\000\000\340\325\177\353\377\177\000\000\020x&\001\000\000\000\000\360\325\177\353\377\177\000\000`\326\177
\353\377\177\000", certificate = {type = 0 '\000', length = 32726}}, m_IdentHash = {{m_Buf = "*x&\001\000\000\000\000@\326\177\353\377\177\000\000`\326\177\353\377\177\000\000 \326\177\353\377\177\000", ll = {19298346, 140737144411712, 
        140737144411744, 140737144411680}}}, m_Verifier = std::unique_ptr<kovri::core::Verifier> containing 0x1267844 <std::unique_ptr<kovri::core::SSUHeader, std::default_delete<kovri::core::SSUHeader> >::get() const+24>, 
  m_ExtendedLen = 21025256, m_ExtendedBuffer = std::unique_ptr<unsigned char> containing 0x7fffeb7fd660 " \376\001\334\377\177"}
(gdb)
```

Problem packet example:
```
(gdb) f 2
#2  0x0000000001290f8f in kovri::core::SSUSessionConfirmedPacket::SetRemoteRouterIdentity (this=0x7fffdd532230, identity=...) at /home/anonimal/kovri/src/core/router/transports/ssu/packet.cc:290
290       m_RemoteIdentity = identity;
(gdb) ins identity
$3 = (const kovri::core::IdentityEx &) @0x7fffeb7fd460: {m_StandardIdentity = {
    public_key = "\200T��\212*i\r�\204\066�P8���)\023 \211�\220\f���\fC���i\213q�Q\b3%\222\225�oa\\n`slZ\236!\004\001�j��\037(�h�\026D�\230\225$�i�%\020zW\217�$\030c�IY{��\211\v֣FP\005\\@�\026����W'7,.K�m\230]hq\220-V\230)n\210\065p���b\0
22��\202!�\t\n=�O\222�٪\032ѯ�\025���u\"T9j�\211u�\225���]\035\221f\202�\221�zYD\237��[�\206��F��=1\207�&F\033\023�h\027S+�"...,
    signing_key = "ZB!\020\234Vi�:\220\237zH\022\067�\213;\031�\034\234k\v\220F��P\016&���\v�tk\235k��\006�Cy��\235���\235i���\216p\224�\212�#�T�\035�h\001\020\033\227\th�ũB�K9\031�\234�˾3\006��\227�J��If}�A�/���3�V�l\a\034�}�\v�`\005\21222\221�\232\213�", certificate = {type = 112 'p', length = 21288}}, m_IdentHash = {{m_Buf = '\000' <repeats 31 times>, ll = {0, 0, 0, 0}}}, m_Verifier = std::unique_ptr<kovri::core::Verifier> containing 0x0, m_ExtendedLen = 10323,
  m_ExtendedBuffer = std::unique_ptr<unsigned char> containing 0x0}

-----

#3  0x00000000012921f7 in kovri::core::SSUPacketParser::ParseSessionConfirmed (this=0x7fffeb7fd770) at /home/anonimal/kovri/src/core/router/transports/ssu/packet.cc:770
        init_length = 1362
        packet = std::unique_ptr<kovri::core::SSUSessionConfirmedPacket> containing 0x7fffdd532230
        identity_size = 724
        identity = {m_StandardIdentity = {
            public_key = "\200T��\212*i\r�\204\066�P8���)\023 \211�\220\f���\fC���i\213q�Q\b3%\222\225�oa\\n`slZ\236!\004\001�j��\037(�h�\026D�\230\225$�i�%\020zW\217�$\030c�IY{��\211\v֣FP\005\\@�\026����W'7,.K�m\230]hq\220-V\230)n\210\06
5p���b\022��\202!�\t\n=�O\222�٪\032ѯ�\025���u\"T9j�\211u�\225���]\035\221f\202�\221�zYD\237��[�\206��F��=1\207�&F\033\023�h\027S+�"...,
            signing_key = "ZB!\020\234Vi�:\220\237zH\022\067�\213;\031�\034\234k\v\220F��P\016&���\v�tk\235k��\006�Cy��\235���\235i���\216p\224�\212�#�T�\035�h\001\020\033\227\th�ũB�K9\031�\234�˾3\006��\227�J��If}�A�/���3�V�l\a\034�}�\v�
`\005\212\221�\232\213�", certificate = {type = 112 'p', length = 21288}}, m_IdentHash = {{m_Buf = '\000' <repeats 31 times>, ll = {0, 0, 0, 0}}}, m_Verifier = std::unique_ptr<kovri::core::Verifier> containing 0x0, m_ExtendedLen = 10323,
          m_ExtendedBuffer = std::unique_ptr<unsigned char> containing 0x0}
        padding_size = 14844381752368197772
```

Inspection (to clarify null source copy):
```
(gdb) f 1
#1  0x0000000001178524 in kovri::core::IdentityEx::operator= (this=0x7fffdd532248, other=...) at /home/anonimal/kovri/src/core/router/identity.cc:229
229           memcpy(m_ExtendedBuffer.get(), other.m_ExtendedBuffer.get(), m_ExtendedLen);
(gdb) ins m_ExtendedBuffer
$15 = std::unique_ptr<unsigned char> containing 0x7fffdd549cc0 ""
(gdb) ins other.m_ExtendedBuffer
$16 = std::unique_ptr<unsigned char> containing 0x0
(gdb) ins m_ExtendedLen
$17 = 10323
(gdb)
```

Cleaning up and catching bad data during copy assignment should be trivial - but I would need more time with the [SSU refactor](https://github.com/monero-project/kovri/issues/140) to pinpoint exactly why it's bad to begin with.

Note: in problem packet, that `padding_size` looks awfully large, so also check for alignment elsewhere.

Heavily referencing #366.

**Comment by anonimal** [comment]
#474 patches to ensure run-time functionality but the real issue appears to be SSU related. I'll relabel this issue as appropriate and reduce the priority since the *worst* of the original issue has been resolved.

**Comment by anonimal** [comment]
After speaking with @danrmiller, investigate the following VM NAT scenario: a host i2p instance attempts to connect to a peer with the same external IP (different port) but routes internally to the guest i2p instance (or vice versa).

Java I2P has a ban list in place for private subnets but kovri currently does not (#473). Even with a list in place, wouldn't routing tables subvert java i2p's banning implementation (without ip-level blocking)? I don't know, research needed.

**Comment by anonimal** [comment]
NOTICE: THIS ISSUE HAS BEEN MOVED TO GitLab. Please continue the discussion there. See https://github.com/monero-project/kovri/issues/1013 for details.
